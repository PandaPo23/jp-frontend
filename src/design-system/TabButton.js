import React from 'react';

import targetedProps from '../utils/targetedProps';
import Button from './Button';

const defaults = {
  borderRadius: 'top.sm',
  borderColor: 'transparent',
  borderTop: 0,
  borderBottom: 4,
  borderLeft: 0,
  borderRight: 0,
  boxShadow: 0,
  bold: false,
  uppercase: true,
  bg: 'surface',
  px: 3,
  py: 3,
  minWidth: 'auto',
  color: 'misc.muted',
  letterSpacing: 'tracked',
};

const activeDefaults = {
  borderColor: 'secondary',
  bold: true,
  color: 'on.background',
};

const TabButton = ({ active = false, children, ...props }) => {
  const activeProps = { ...activeDefaults, ...targetedProps(props, 'active') };
  return (
    <Button {...defaults} {...(active && activeProps)} {...props}>
      {children}
    </Button>
  );
};

export default TabButton;
