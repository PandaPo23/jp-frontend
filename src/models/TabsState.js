import { action, computed, decorate, observable } from 'mobx';
import get from 'lodash/get';

export default class TabsState {
  _tabs = [];
  activeTab = '';

  constructor(tabs, activeTabId) {
    this._tabs = tabs;
    this.activeTab = activeTabId || get(tabs, '[0].id') || '';
  }

  setActiveTab(id) {
    this.activeTab = id;
  }

  get tabs() {
    return this._tabs.map(({ id, name }) => ({
      id,
      name,
      active: this.activeTab === id,
    }));
  }

  get activeTabName() {
    return (
      get(this._tabs.find((tab) => tab.id === this.activeTabId), 'name') || ''
    );
  }
}

decorate(TabsState, {
  activeTab: observable,
  setActiveTab: action,
  activeTabName: computed,
  tabs: computed,
});
