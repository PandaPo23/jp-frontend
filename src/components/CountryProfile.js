import React from 'react';
import { inject, observer } from 'mobx-react';
import { withTranslation } from 'react-i18next';
import compose from 'lodash/fp/compose';
import PropTypes from 'prop-types';

import { FlexPanel, ImageSorter, TextContent } from '../design-system';

export const CountryProfile = ({ country, images, mapIsInitialized }) =>
  country ? (
    <FlexPanel bg="background" fontSize={4} flex={1}>
      {mapIsInitialized && (
        <ImageSorter
          height={300}
          minHeight={300}
          borderRadius="none"
          images={images}
        />
      )}
      <TextContent py={0} text={country.description} />
    </FlexPanel>
  ) : null;

CountryProfile.propTypes = {
  country: PropTypes.object,
  t: PropTypes.func.isRequired,
};

export default compose(
  withTranslation('common'),
  inject(({ app }, { countryCode }) => ({
    country: app.countryInfo,
    images: app.getCountryImages(countryCode),
    mapIsInitialized: app.map.initialized,
  })),
  observer
)(CountryProfile);
