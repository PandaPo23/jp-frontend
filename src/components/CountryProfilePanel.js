import React from 'react';
import { inject, observer, PropTypes as MobxPropTypes } from 'mobx-react';
import { withRouter } from 'react-router-dom';
import compose from 'lodash/fp/compose';
import PropTypes from 'prop-types';

import { FlexPanel, IconButton, SlidingPanel } from '../design-system';
import { withTranslation } from 'react-i18next';
import { CountryProfile, Tabs } from '.';

export const CountryProfilePanel = (props) => {
  const {
    countryTabsState,
    isOpen,
    onClose,
    t,
    setCountry,
    width = 600,
    ...otherProps
  } = props;
  return (
    <SlidingPanel
      id="country-profile"
      maxWidth={width}
      closeType="none"
      open={isOpen}
      onToggle={onClose}
      {...otherProps}
    >
      <FlexPanel
        bg="background"
        flex={1}
        alignContent="flex-start"
        borderLeft={1}
        width={width}
      >
        <FlexPanel fontSize={4} bg="surface">
          <IconButton
            p={3}
            bg="surface"
            onClick={onClose}
            name="arrow-back"
            mb={1}
            labelFontSize={5}
            labelMl={2}
            labelColor="on.background"
          >
            {t('countries_visited', 'Countries Visited')}
          </IconButton>
          <Tabs
            tabs={countryTabsState}
            shouldTranslate={false}
            onChange={setCountry}
          />
        </FlexPanel>
        <CountryProfile countryCode={countryTabsState.activeTab} />
      </FlexPanel>
    </SlidingPanel>
  );
};

CountryProfilePanel.propTypes = {
  countryTabsState: MobxPropTypes.objectOrObservableObject.isRequired,
  t: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
};

export default compose(
  withRouter,
  withTranslation('common'),
  inject(({ app }, { history }) => ({
    isOpen: app.isCountryProfilePanelOpen,
    onClose: () => history.push(`/trips/${app.currentTrip.id}`),
    countryTabsState: app.route.countryTabsState,
    width: app.countryPanelWidth,
    setCountry: (code) =>
      app.goto(`/trips/${app.currentTrip.id}/country/${code}`),
  })),
  observer
)(CountryProfilePanel);
