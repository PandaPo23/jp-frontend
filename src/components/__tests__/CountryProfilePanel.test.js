import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CountryProfilePanel } from '../CountryProfilePanel';
import CountryProfile from '../CountryProfile';
import TabButton from '../../design-system/TabButton';
import TabsState from '../../models/TabsState';

jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
  Map: () => ({}),
}));

const countryTabsState = new TabsState(
  [{ id: 'JP', name: 'Japan' }, { id: 'TH', name: 'Thailand' }],
  'TH'
);

describe('CountryProfilePanel component', () => {
  it('renders correctly', () => {
    const localProps = {
      isOpen: true,
      onClose: () => {},
      countryTabsState,
      t: (k, v) => v,
    };
    const wrapper = shallow(<CountryProfilePanel {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(CountryProfile).prop('countryCode')).toEqual('TH');
  });

  it('switches between country tabs when tab buttons are clicked', () => {
    const localProps = {
      isOpen: true,
      onClose: () => {},
      countryTabsState,
      t: (k, v) => v,
    };
    const wrapper = shallow(<CountryProfilePanel {...localProps} />);
    expect(countryTabsState.activeTab).toEqual('TH');
    setTimeout(() => {
      wrapper
        .find(TabButton)
        .first()
        .simulate('click');
    }, 1000);
    expect(countryTabsState.activeTab).toEqual('TH');
  });
});
