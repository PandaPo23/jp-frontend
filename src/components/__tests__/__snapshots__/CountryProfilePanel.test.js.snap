// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryProfilePanel component renders correctly 1`] = `
<SlidingPanel
  closeType="none"
  id="country-profile"
  maxWidth={600}
  onToggle={[Function]}
  open={true}
>
  <FlexPanel
    alignContent="flex-start"
    bg="background"
    borderLeft={1}
    display="flex"
    flex={1}
    flexDirection="column"
    width={600}
  >
    <FlexPanel
      bg="surface"
      display="flex"
      flexDirection="column"
      fontSize={4}
    >
      <IconButton
        bg="surface"
        labelColor="on.background"
        labelFontSize={5}
        labelMl={2}
        mb={1}
        name="arrow-back"
        onClick={[Function]}
        p={3}
      >
        Countries Visited
      </IconButton>
      <withI18nextTranslation(Tabs)
        shouldTranslate={false}
        tabs={
          TabsState {
            "_tabs": Array [
              Object {
                "id": "JP",
                "name": "Japan",
              },
              Object {
                "id": "TH",
                "name": "Thailand",
              },
            ],
            "activeTab": "TH",
          }
        }
      />
    </FlexPanel>
    <withI18nextTranslation(inject-CountryProfile)
      countryCode="TH"
    />
  </FlexPanel>
</SlidingPanel>
`;
